### DISCUSSION TITLE
How to templating similar ImagePolicy & ImageUpdateAutomation

### DISCUSSION THREAD

**Discussion by belkevglaz** [fluxcd/flux2#3834] [General] [answered]
I'd like to use Image Automation. I have created crds ImageRepository, ImageUpdateAutomation and ImagePolicy for the application.
For one app it works fine, but for dozens of apps I have to create dozens of almost identical manifests.
Is there a way to customise or template all these manifests?

**Comment by kingdonb** [comment, marked as answer]
There are several solutions for this problem that I can offer you, and some of them are part of solutions that are sold by Weaveworks

* GitOpsSets
* GitOpsTemplates
* Kluctl Templates

This is exactly what solutions like these templates are for. A more basic version of the GitOpsTemplates is called KustomizationSet, it's provided by this archived open source project: https://github.com/gitops-tools/kustomization-set-controller

You can probably get a better idea of how this can be used in anger from the Weave GitOps Enterprise docs:

https://docs.gitops.weave.works/docs/gitops-templates/intro/

Kluctl has their own version here:

https://kluctl.io/docs/template-controller/

I think that most of the capabilities are available in Kluctl, I'm not sure though because I am not using it.

Here's a nice feature that will give you an idea of what differentiates these tools:
https://docs.gitops.weave.works/docs/gitops-templates/supported-langs/

This is for problems caused when YAML parsers are consuming your quotes, for example, or when you need something more involved than you can do without a tool like gotl (helm's template engine). You can also accomplish most of this with a Kustomize base and patching. It's not a perfect solution, which is why we offer templating as well.

The main problems you'll run into around for example copying secrets or promoting pull requests from one environment to another, and these are the type of problems you'll find solved in more detail by our enterprise solution. Kluctl tackles some of that as well.

Hope this helps!

**Reply by kingdonb** [reply]
Maybe we can add an updated doc that shows image update automation being templated across several tenants in a reasonable way using only Flux's tools? I'll leave this open for now in case anyone can provide this example as a reference, so we can convert this to an issue and potentially improve our documentation with that.
